File: лаба2.cpp
//мультисписок на С++
#include <iostream>
#include <vector>
#include <memory>
class Node {
public:
    std::string value;
    std::vector<std::shared_ptr<Node>> children;
    Node(const std::string& val) : value(val) {}
};
int main() {
    auto root = std::make_shared<Node>("Root");
    root->children.push_back(std::make_shared<Node>("Child 1"));
    root->children.push_back(std::make_shared<Node>("Child 2"));
    // Отображение
    std::cout << root->value << std::endl;
    for (auto& child : root->children) {
        std::cout << "  " << child->value << std::endl;
    }
    return 0;
}

//очередь на С++
#include <iostream>
#include <queue>
int main() {
    std::queue<int> queue;
    queue.push(1);  // Добавление элемента
    queue.push(2);
    std::cout << queue.front() << std::endl;  // Вывод первого элемента (1)
    queue.pop();  // Удаление первого элемента
    return 0;
}

//дек на С++
#include <iostream>
#include <deque>
int main() {
    std::deque<int> deque;
    deque.push_back(1);  // Добавление в конец
    deque.push_front(2); // Добавление в начало
    std::cout << deque.back() << std::endl;  // Вывод последнего элемента (1)
    deque.pop_back();  // Удаление последнего элемента
    return 0;
}

//очередь на С++
#include <iostream>
#include <queue>
int main() {
    std::priority_queue<std::pair<int, std::string>> pq;
    pq.push({1, "task 1"});
    pq.push({0, "task 2"});
    std::cout << pq.top().second << std::endl;  // Вывод задачи с наивысшим приоритетом (task 2)
    pq.pop();  // Удаление элемента с наивысшим приоритетом
    return 0;
}

File: лаба2.py
#пример использования мультисписка
class Node:
    def __init__(self, value):
        self.value = value
        self.children = []
class MultiList:
    def __init__(self):
        self.root = None
    def add_child(self, parent, child):
        parent.children.append(child)
# Пример использования
root = Node("Root")
child1 = Node("Child 1")
child2 = Node("Child 2")
multi_list = MultiList()
multi_list.root = root
multi_list.add_child(root, child1)
multi_list.add_child(root, child2)

#пример использования очереди
rom collections import deque
queue = deque()
queue.append(1)  # Добавление элемента
queue.append(2)
print(queue.popleft())  # Удаление и вывод первого элемента

#пример использования дек
from collections import deque
deque_example = deque()
deque_example.append(1)  # Добавление в конец
deque_example.appendleft(2)  # Добавление в начало
print(deque_example.pop())  # Удаление и вывод последнего элемента

#пример использования приоритетной очереди
import heapq
priority_queue = []
heapq.heappush(priority_queue, (1, 'task 1'))  # (приоритет, задача)
heapq.heappush(priority_queue, (0, 'task 2'))
print(heapq.heappop(priority_queue))  # Удаление и вывод элемента

